actors/market/src/lib.rs

```diff
@@ -543,137 +550,226 @@ impl Actor {
-            let mut sector_deals: Vec<(SectorNumber, SectorDealIDs)> = vec![];
-
-            for p in params.sectors {
-                let proposal_array = st.get_proposal_array(rt.store())?;
-
-                if p.deal_ids.iter().any(|id| activated_deals.contains(id)) {
-                    log::warn!(
-                        "failed to activate sector containing duplicate deals {:?}",
-                        p.deal_ids
-                    );
-                    batch_gen.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
-                    continue;
-                }
+            let mut sectors_deals: Vec<(SectorNumber, SectorDealIDs)> = vec![];
 
-                let proposals = match get_proposals(&proposal_array, &p.deal_ids, st.next_id) {
-                    Ok(proposals) => proposals,
-                    Err(e) => {
-                        log::warn!("failed to get proposals for deals {:?}: {:?}", p.deal_ids, e);
-                        batch_gen.add_fail(e.exit_code());
-                        continue;
+            'sector: for p in params.sectors {
+                let mut verified_infos = Vec::new();
+                let mut nonverified_deal_space: BigInt = BigInt::zero();
+                for deal_id in &p.deal_ids {
+                    // Check each deal is present only once, within and across sectors.
+                    if activated_deals.contains(deal_id) {
+                        log::warn!("failed to activate sector, duplicated deal {}", deal_id);
+                        batch_gen.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
+                        continue 'sector;
                     }
-                };
 
-                let deal_spaces = match validate_and_return_deal_space(
-                    &proposals,
-                    &miner_addr,
-                    p.sector_expiry,
-                    curr_epoch,
-                    None,
-                ) {
-                    Ok(ds) => ds,
-                    Err(e) => {
-                        log::warn!("failed validate deals {:?}: {}", p.deal_ids, e);
-                        batch_gen.add_fail(e.exit_code());
-                        continue;
+                    let proposal = match preactivate_deal(
+                        rt,
+                        *deal_id,
+                        &proposals,
+                        &states,
+                        &pending_deals,
+                        &miner_addr,
+                        p.sector_expiry,
+                        curr_epoch,
+                        st.next_id,
+                    )? {
+                        Ok(v) => v,
+                        Err(e) => {
+                            log::warn!("failed to activate deal: {}", e);
+                            batch_gen.add_fail(e.exit_code());
+                            continue 'sector;
+                        }
+                    };
+
+                    // No continue below here, to ensure state changes are consistent.
```

Comment:
I don't understand this because I can't find any state changes above in `preactivate_deal` and therefore it feels like we can continue to the next sector on error.  What state change am I missing above?

Reply:
Ah ok is it correct that the problem scenario 1. multiple deals per sector 2. one deal has already been activated succesfully 3. a later deal fails and therefore we can't easily roll back the earlier deal state changes?

Reply:
In that case won't *all* failures in the deal loop also cause the same problem when there are more than one deal ids so we should be either a) failing the whole thing on any deal error (preactivate_deal or duplicate deals) or b) figuring out how to do atomic rollback at the sector level?

Reply:
I think atomic rollback would be annoying but nowhere near as hard as the case where we call other actors.  We could just keep inner accumulators of verified_infos unverified_deal_space and activated deal states and add them to the main accumulator at the end of the deal ids loop

Reply:
Thanks, I had overlooked this last bit when sending for review. I've cleaned it up now to get atomic failure of sectors.